graph/src/blockchain/block_stream.rs

```diff
@@ -24,22 +60,729 @@ pub trait TriggersAdapter<C: Blockchain> {
     // `step_size` is the suggested number blocks to be scanned.
     async fn scan_triggers(
         &self,
-        chain_base: BlockPtr,
-        step_size: u32,
+        from: BlockNumber,
+        to: BlockNumber,
         filter: C::TriggerFilter,
-    ) -> Result<Vec<BlockWithTriggers<C>>, ScanTriggersError>;
+    ) -> Result<Vec<BlockWithTriggers<C>>, Error>;
```

Comment:
But that interface would be wrong (or expensive) for Ethereum: if the caller makes no guarantee about finality, the adapter has to check that themselves. Right now, this is only called for blocks that are known to be final, and I should have added a comment to that effect. I just think that getting this right will be a lot easier with another example in hand. I think what you are getting at is who needs to know about finality: only the chain itself, or the generic framework, too. Right now, that knowledge is split, since the block stream depends on it. If we want to remove this knowledge from the framework, we'd have to consider the `BlockStream` chain-specific, which AFAICT would obviate the need for things like a `TriggerAdapter`.

Reply:
Fair point that the Ethereum implementation wouldn't really be correct, so we can't have a single interface that's correct and performant for both cases. Agreed that we can figure this out if we start to work on a second block stream implementation.